addons/mail/models/mail_message_reaction.py

```diff
@@ -11,7 +11,7 @@ class MailMessageReaction(models.Model):
     _order = 'id desc'
     _log_access = False
 
-    message_id = fields.Many2one(string="Message", comodel_name='mail.message', ondelete='cascade', required=True, readonly=True)
+    message_id = fields.Many2one(string="Message", comodel_name='mail.message', ondelete='cascade', required=True, readonly=True, index=True)
```

Comment:
Isn't it already indexed through `_partner_unique` or `_guest_unique`?

Reply:
These are partial unique constraints, for example for `_partner_unique`, for rows where `partner_id` are NULL, the index cannot cover index scans that would hit the first key `message_id`. Same argument for the other index.

Reply:
I see, feel free to ignore my comment about this one then and keep the new index, I didn't really know how it would work in this case as it is a bit tricky. Both index combined cover the full range of message_id but I guess it can't really combine them like that.

Reply:
> Both index combined cover the full range of message_id but I guess it can't really combine them like that

Wait they are both complementary? Wasn't aware of that, then if that's the case:
- if `partner_id IS NULL` -> the guest is set, therefor it can use `_guest_unique`
- if `guest_id IS NULL` -> the partner is set, therefor it can use `_partner_unique`

Let me know if didn't misunderstand you, and I'll drop this index.

Reply:
Yes if you check `partner_or_guest_exists` there be must one of `partner_id` or `guest_id` so all `message_id` are covered if both separate unique are taken together.

Reply:
I think the problem is it is partially indexed.
You have to specify the `partial index condition` to ask postgresql use it.
e.g. `model.search([('message_id', '=', 100), ('content', '=', 'xxx'), ('partner_id', '=', 200)])` cannot use the `_partner_unique`
you have to search `model.search([('message_id', '=', 100), ('content', '=', 'xxx'), ('partner_id', '=', 200), ('guest_id', '=', False)])`

Reply:
Thanks for pointing this out @HydrionBurst, I was just curious the see if with the constraint present, postgres would be able to deduce enough, but after a quick test, alas, it's not the case. As you've said, if you don't specify the partial clause of the index in the query where clause, it won't use any of the indexes, even if there is a constraint that enforces that both partial indexes should contains the totality of the table's keys.

So sadly, the addition a the non-partial index on message_id is necessary, even if it seemed covered by the two unique indexes.